Replicator/DBAccess.cc

```diff
@@ -305,8 +305,8 @@ namespace litecore { namespace repl {
             lock_guard<mutex> lock(_tempSharedKeysMutex);
             if (!_tempSharedKeys || _tempSharedKeysInitialCount < dbsk.count()) {
                 // Copy database's sharedKeys:
-                _tempSharedKeys = SharedKeys::create(dbsk.stateData());
                 _tempSharedKeysInitialCount = dbsk.count();
```

Comment:
Are you sure you don't want to put the loop here?  I was much persuaded by your argument that you really didn't have a clear root cause, and that it was unclear why this tweaky change fixed it.

Reply:
That is, the original first statement relies on the side-effect of second statement that reads the count. Reading the if statement, dbsk.count() may not be called if !_tempSharedKey == true. I wonder if the problem could also be fixed by swapping the two expressions in the if statement.

Reply:
Will getting the count from the `_tempSharedKeys` also solve the issue? I think doing that will eliminate the atomic issue b/w creating a temp shared key and getting its initial count?

Reply:
From the ShredKeys.cc in fleece, there is a mutex lock to guard `_count`. It seems like there is one place doesn't do that and it's in the `writeState()` used by `stateData()` method. I don't know if this contributes to the issue or not but it's a suspicous.

https://github.com/couchbase/fleece/blob/dcaa54943e6fcf05a610476912299873b5328bf7/Fleece/Core/SharedKeys.cc#L91

Reply:
@bmeike I suppose I could add it anyway, but this change made just as many passes as the loop did.

@pasin No, in fact that is the core of the problem.  It is getting created incorrectly, so from the temp shared keys point of view the count legitimately is zero. I am going to try using the mutex in stateData as well.